server/base/src/main/java/org/apache/accumulo/server/metadata/ConditionalTabletMutatorImpl.java

```diff
@@ -154,6 +155,16 @@ public Ample.ConditionalTabletMutator requireAbsentOperation() {
     return this;
   }
 
+  @Override
+  public Ample.ConditionalTabletMutator requireAbsentNextOperation() {
+    Preconditions.checkState(updatesEnabled, "Cannot make updates after calling mutate.");
+    Condition c =
```

Comment:
With  next operation and operation ids, would only expect users of Ample to do the following and nothing else.
  
1. Make tablet mutations only IF no operation id is set
2. Make tablet mutations only IF a particular operation id is set
3. Set the next operation id only IF no operation id is set AND no location is set
4. Set the operation id and delete the next operation id only IF the next operation id being deleted is the same as the operation id we are setting AND no location is set
5. Set the operation id only IF no operation id is set AND no next operation id is set AND no location is set

Wonder if we can design the ample API such that its only possible to do those things.  The way the API is now other things could be done that are not on the list above, but those are not expected (like setting a next opid and opid).

Wondering if we can have top level ample methods that do 3,4, and 5 in the list above.  With those may not even need method on tablet mutator to set the next operation id or set the operation id.  I will open a follow on issue about this, its not something that needs to be done in this PR.

Reply:
> 3. Set the next operation id only IF no operation id is set AND no location is set

I think that this could cause "flapping" where tablets are in a hosting / unload loop. I think the nextOperationId could be set regardless of the value of operationId. 

> 4. Set the operation id and delete the next operation id only IF the next operation id being deleted is the same as the operation id we are setting AND no location is set

I didn't make the assumption that the `next` and `opid` id values were the same. I took the approach that I just needed something in the `next` column to trigger an unhost and then deleted the `next` value in `TabletMutatorBase.startOperation` when it sets the `opid`. Something in `next` causes the tablet to be unhosted. When the tablet no longer has a location and no current `opid`, then a Fate transaction can run. I'm not sure it has to be the same one as there could be multiple queued up. That was my thought process anyway.

Reply:
>  I'm not sure it has to be the same one as there could be multiple queued up. That was my thought process anyway.

I had thought we would want the ability to have multiple operations queues up, but looking at this PR I realized that may not make sense.  What I realized is that when anything sets an opid that its a destructive operation and that the tablet will no longer exists after it runs.  So if a split operation is trying to split a specific tablet and it sees something else has an opid set on the tablet its trying to split, then there is no need for it to continue because the tablet its trying  to split is in the process of being deleted (and its data possibly merged or split into other tablets).  So in this scenario it does not make sense for split operation to queue itself up and then later find its tablet no longer exists.

For the situation where a split is trying to start and it sees a location and a next op id (that is not itself), what should it do? 

 1. It could just abort, assuming the other operation will eventually get the opid.  
 2. It could wait and see if the other operation gets the next opid (without making any changes to tablet metadata).  
 3. It could queue itself up for getting the opid.  The current metadata schema does not support queuing because the opid is in the value, would need to be in the qualifier to support queuing.

I am not sure what is best.  Kinda leaning towards 1 or 2 because they seem simpler to implement, maintain, and understand and also seem correct.  Also I am unable to think of benefits for option 3 over 1 and 2, and it seems more complicated to implement and maintain.

If pursuing option 1 or 2, then we do not have to check the next opid and opid are the same for correctness (at least for the cases I can think of).  However I like the idea of checking it if its what we expect.  If there is no queuing and no other operation will try to get the opid when next opid is set, then checking it adds validates assumptions.  But if its not the expected behavior, then there is no need to check it.

Reply:
> If multiple operations are not going to queue for the tablet, then I'm not sure we need a nextOperationId. We could achieve the same thing by divorcing the setting of the operationId and checking of the location. 

I really really like that!  Just need to ensure that every operation properly waits for the location to be unset.  That should be easy to do, there are only three destructive tablet operations.  Should be easy to ensure they follow the protocol.  Also, every conditional mutation made by a destructive operation could require an absent location an extra validation.  We could build this into ample, when an update requires an operation id it could also automatically require an absent location.

Reply:
Ok. I'm going to back out the nextOperationId changes in this PR and make the necessary modifications.

Reply:
> Also, every conditional mutation made by a destructive operation could require as an absent location an extra validation. We could build this into ample, when an update requires an operation id it could also automatically require an absent location.

Not sure you can do this if the existence of an operationId is going to trigger an unload.

Reply:
> If multiple operations are not going to queue for the tablet, then I'm not sure we need a nextOperationId. 

A lot of the current code assumes that opid and location are never set at the same time.  So we will need to track those assumptions down and change them to the new assumption.  Should be able to look for code that requires an operation id in a conditional update and see what its doing with locations.  Also would need to change the tablet metadata code, it may throw an exception if both are set.

Reply:
> Not sure you can do this if the existence of an operationId is going to trigger an unload.

Thinking the following could happen for a split operation.

 1. Attempts to set the operation id, if can't set it aborts.
 2. After setting the operation id, it waits for the location to be unset
 3. Any changes to the tablet metadata made by the split operation will set conditions to require their operation id and and an absent location.  The location is expected to be absent because no tablet should load if the opid is set and we waited for it be absent in step 2.  So setting the condition to check for an absent location is just an extra validation at this point.